### PR TITLE
Aws instance profile warning

### PR DESCRIPTION
**Pull request by tlm**
With the introduction of instance profile code there is a potentially error for users when destroying controllers that have a very limited permission set in use because 2.9 now requires more permissions then it did previously. This PR introduces the new checks as warning's should this case arise.

## Checklist

 - [x] Requires a [pylibjuju](https://github.com/juju/python-libjuju) change
 - [x] Added [integration tests](https://github.com/juju/juju/tree/develop/tests) for the PR
 - [x] Added or updated [doc.go](https://discourse.jujucharms.com/t/readme-in-packages/451) related to packages changed
 - [x] Comments answer the question of why design decisions were made

## QA steps

Create a aws credential user with the policy found [here](https://discourse.charmhub.io/t/juju-aws-permissions/5307). From this policy remove the `"iam:ListInstanceProfiles"` and `"iam:ListRoles"` actions.

Add the newly created AWS user to your local Juju credentials and bootstrap a controller.

Kill the controller with:
`juju kill-controller -t0 -y`

You should see warnings in the output about the lack of permissions and manual cleanup may be required.

## Documentation changes

N/A

## Bug reference

https://bugs.launchpad.net/juju/+bug/1954915

### CODE REVIEW

File: provider/ec2/environ.go [hunk 1]
```diff
@@ -1890,7 +1892,9 @@ func (e *environ) destroyControllerManagedModels(ctx context.ProviderCallContext
 	}
 
 	roles, err := listRolesForController(ctx, e.iamClient, controllerUUID)
-	if err != nil {
+	if errors.IsUnauthorized(err) {
+		logger.Warningf("unable to roles for deletion, roles may have to be manually cleaned up for controller %q", controllerUUID)
```
Comment (wallyworld) [inline provider/ec2/environ.go:1896]: typo
also, can we include (if available) the controller name in the message (here and above)
eg
`... manually cleaned up for controller "foo" (uuid-dead-beef)`

File: provider/ec2/iam.go [hunk 2]
```diff
@@ -359,7 +359,11 @@ func listInstanceProfilesForController(
 		})
 
 		if err != nil {
-			return nil, errors.Annotatef(
+			var opHTTPErr *awshttp.ResponseError
+			if stderrors.As(err, &opHTTPErr) && opHTTPErr.HTTPStatusCode() == http.StatusForbidden {
```
Comment (wallyworld) [inline provider/ec2/iam.go:363]: we have this here and another below, and we also have
`opHTTPErr.HTTPStatusCode() == http.StatusNotFound` in 2 places
can we introduce 2 helper methods and use those
`isNotFoundError(err error) bool`
`isUnauthorisedError(err error) bool`

File: provider/ec2/environ.go (hunk 1 above)
Comment (tlm) [reply provider/ec2/environ.go:1896]: we don't have the controller name.

File: provider/ec2/internal/testing/iam_server.go [hunk 3]
```diff
@@ -29,6 +30,12 @@ func NewIAMServer() (*IAMServer, error) {
 	return srv, nil
 }
 
+func (i *IAMServer) ProducePermissionError(p bool) {
+	i.mu.Lock()
+	defer i.mu.Unlock()
+	i.producePermissionError = p
+}
+
```
Comment (wallyworld) [inline provider/ec2/internal/testing/iam_server.go:38]: Yuck :-)
Can we instead have it return a permission error based on something like the prefix being used? Or the id value or something we can set as part of the role or profile?

**Review by wallyworld — APPROVED** [review]
Just a few small cleanups

File: provider/ec2/iam.go (hunk 2 above)
Comment (tlm) [reply provider/ec2/iam.go:363]: Fixed

File: provider/ec2/internal/testing/iam_server.go (hunk 3 above)
Comment (tlm) [reply provider/ec2/internal/testing/iam_server.go:38]: Not really

**Pull request closed without merging**